gapis/api/templates/vulkan_gfx_api_extras.tmpl

```diff
@@ -667,6 +667,45 @@ bool Vulkan::replayCreateVkDebugReportCallback(Stack* stack, bool pushReturn) {
     }
 }
 ¶
+bool Vulkan::replayWaitForFences(Stack* stack, bool pushReturn) {
+   auto expecteds = stack->pop<uint64_t*>();
+   auto fences = stack->pop<uint64_t*>();
+   auto count = stack->pop<uint64_t>();
+   auto device = static_cast<size_val>(stack->pop<size_val>());
+   if (stack->isValid()) {
+        GAPID_DEBUG("vkWaitForFences(%zu)", device);
+        uint32_t fence_status = VK_TIMEOUT;
+        for (uint64_t i = 0; i < count; ++i) {
+          auto expected = expecteds[i];
+          if (expected == VK_NOT_READY) {
+            continue;
+          }
+
+          if (expected == VK_ERROR_DEVICE_LOST) {
+            fence_status = VK_ERROR_DEVICE_LOST;
+            continue;
+          }
+
+          do {
+            fence_status = mVkDeviceFunctionStubs[device].vkGetFenceStatus(device, fences[i]);
+          } while (fence_status == VK_NOT_READY);
```

Comment:
Busy wait is probably better. We want to wake up ASAP when the gpu I'd done.  Burning a core is totally fine as we are only using one, and unnecessary waits may throw off you timing

Reply:
Why are we not just calling WaitForFences?

I recall GAPID preferred spinning on GetFenceStatus as a workaround for some old broken driver (@AWoloszyn which?), but that driver should not be in the supported set for AGI.

If we decide we must perpetuate the workaround, there should at least be a comment here.

Reply:
From what I understood from my conversion with Andrew that we do not want to change, alter or effect the original call to the driver. Otherwise we may hide driver bugs. But I think the best answer can be given by him.

Reply:
+1 for a comment here to clearly explain why we do this. Let me try to draft one, and that will also let me know if I understand all this correctly:

```
We cannot simply replay vkWaitForFences because of timeout subtleties.
Consider this call: vkWaitForFences(..., false, 123456)
It's waiting on any fence (waitAll is false) with a given timeout. If this call
returns VK_SUCCESS at capture time, it means that one fence has signaled
before the timeout. At replay time, there is no guarantee that the same fence
will signal, or that the timeout will not be hit. So in order to have the replay
mirror the capture-time behavior, we busy-wait on all the fences that were
known to be signaled at capture time.
```

But that makes me realize: should we do this only when we have waitAll == false and return == VK_SUCCESS ?

And on the flip side, if the waitForFence did return VK_TIMEOUT at capture time, there is no guarantee that it will hit the timeout at replay time: should we busy wait for fences status and then sleep if need be to reach the timeout duration before returning VK_TIMEOUT?

By the way, the current code seems incorrect as it propagates the return value of a vkGetFenceStatus as the return value of replayWaitForFences.

Reply:
> From what I understood from my conversion with Andrew that we do not want to change, alter or effect the original call to the driver. 

All of these GetFenceStatus calls were never made by the app. The app called WaitForFences, once.

Reply:
We cannot call WaitForFences once though. If there is a timeout, (or multiple fences) we may not end up with the same set of ready fences.
Particularly if replay returns VK_TIMEOUT,and trace returned VK_SUCCESS.... (even for one fence) .. .what do we set timeout to the second time.... the same as the first?, zero? INF?

Reply:
As for the driver, I cannot remeber which. I actually think it was several different drivers that ran into the problem.

Reply:
If the trace returned VK_SUCCESS then to be correct you should WaitForFences with timeout=inf waitAll=true on the subset of the fences which you observed to be ready. That's effectively what this loop over fences + spinning is handrolling.

Reply:
That should work, so long as all of the drivers are nice and stable. I am not confident that they are. It would require writing applications that exhibit this behavior, and test agains all of the drivers/devices you support. I am not personally confident that every driver you care about is not broken in this way. If they ARE , finding this bug is extremely tricky.

Reply:
I have realised the fenceStatus and I will fix it. It does not change the general logic though.

@chrisforbes I have meant that we cannot change the VkWaitForFences and call for a different set of fences but I think I have misworded.

My first attempt for this one was getting the set of fences that actually signaled and just waiting on them with a VKWaitForFences and I guess this is close to what @chrisforbes suggests but Andrew didn't want to do that because that would change the actual wait operation.

I am happy to change the code to a more correct logic but I think that's not a decision I can do myself. I will wait for the @chrisforbes and @AWoloszyn's input.

Current solution ensures that we have at least waited for the fences that has been signaled, if none of them signaled then we will not wait at all. 

If you have any suggestions on how we should change this, please put them. Otherwise, I am happy to merge after the return value is fixed. I think the logic is correct even though it may not be the optimal.

Reply:
I agree with Chris, that we should just wait on all the fences we expect to be signaled based on the capture. If we want to attempt to also make buggy drivers debuggable, I suggest we do this pseudo-code:
```
vkWaitForFences(<same params as capture>)
if (numFences > 1 && !waitAll) {
  vkWaitForFences(<fences signaled at capture time>, waitAll=true, timeout=MAX_UINT64)
}
```
This way we replay what was captured, but also ensure the state we expect. The only sticky point here is if there is a bug in the driver for the waitAll case, which the app may never use, but I don't think that is worth worrying about.

Reply:
I will reimplement the replay part but I am not sure if we have to call vkWaitForFences twice. I suggest only the

if signaledFences.empty() {
    return timeout
} else {
     return vkWaitForFences(<fences signaled at capture time>, waitAll=true, timeout=MAX_UINT64)
}

Reply:
And update to this: 

instead of `return timeout` `VkWaitForFences(original fences, originalWaitAll, originalTimeout)`